File: 02500/BOJ_2564.py
import sys

input = sys.stdin.readline

w, h = map(int, input().split())
n = int(input())
store = []

for _ in range(n):
    a, b = map(int, input().split())
    store.append((a, b))
x, y = map(int, input().split())

ans = 0

for s in store:
    a, b = s
    
    if x==1:
        if a==1: ans += abs(y-b)
        elif a==2: ans += h + min(y+b, (w-y)+(w-b))
        elif a==3: ans += y + b
        elif a==4: ans += (w-y) + b

    elif x==2:
        if a==1: ans += h + min(y+b, (w-y)+(w-b))
        elif a==2: ans += abs(y-b)
        elif a==3: ans += y + (h-b)
        elif a==4: ans += (w-y) + (h-b)

    elif x==3:
        if a==1: ans += y + b
        elif a==2: ans += (h-y) + b
        elif a==3: ans += abs(y-b)
        elif a==4: ans += w + min(y+b, (h-y)+(h-b))

    elif x==4:
        if a==1: ans += y + (w-b)
        elif a==2: ans += (h-y) + (w-b)
        elif a==3: ans += w + min(y+b, (h-y)+(h-b))
        elif a==4: ans += abs(y-b)

print(ans)

File: 02500/BOJ_2564.java
import java.io.*;
import java.util.*;

class Main {
    private void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        int W = Integer.parseInt(st.nextToken());
        int H = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        ArrayList<ArrayList<Integer>> arr = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            arr.add(new ArrayList<>());
            arr.get(i).add(a);
            arr.get(i).add(b);
        }
        st = new StringTokenizer(br.readLine());
        int x = Integer.parseInt(st.nextToken());
        int y = Integer.parseInt(st.nextToken());
        int ans = 0;
        for (int i = 0; i < N; i++) {
            if (x == 1) {
                if (arr.get(i).get(0) == 1) {
                    ans += Math.abs(y - arr.get(i).get(1));
                } else if (arr.get(i).get(0) == 2) {
                    ans += H + Math.min(y + arr.get(i).get(1), 2*W - (y+arr.get(i).get(1)));
                } else if (arr.get(i).get(0) == 3) {
                    ans += y + arr.get(i).get(1);
                } else if (arr.get(i).get(0) == 4) {
                    ans += W-y + arr.get(i).get(1);
                }
            } else if (x == 2) {
                if (arr.get(i).get(0) == 2) {
                    ans += Math.abs(y - arr.get(i).get(1));
                } else if (arr.get(i).get(0) == 1) {
                    ans += H + Math.min(y + arr.get(i).get(1), 2*W - (y+arr.get(i).get(1)));
                } else if (arr.get(i).get(0) == 3) {
                    ans += y + H-arr.get(i).get(1);
                } else if (arr.get(i).get(0) == 4) {
                    ans += W-y + H-arr.get(i).get(1);
                }
            } else if (x == 3) {
                if (arr.get(i).get(0) == 3) {
                    ans += Math.abs(y - arr.get(i).get(1));
                } else if (arr.get(i).get(0) == 4) {
                    ans += W + Math.min(y + arr.get(i).get(1), 2*H - (y+arr.get(i).get(1)));
                } else if (arr.get(i).get(0) == 1) {
                    ans += y + arr.get(i).get(1);
                } else if (arr.get(i).get(0) == 2) {
                    ans += H-y + arr.get(i).get(1);
                }
            } else if (x == 4) {
                if (arr.get(i).get(0) == 4) {
                    ans += Math.abs(y - arr.get(i).get(1));
                } else if (arr.get(i).get(0) == 3) {
                    ans += W + Math.min(y + arr.get(i).get(1), 2*H - (y+arr.get(i).get(1)));
                } else if (arr.get(i).get(0) == 1) {
                    ans += y + W-arr.get(i).get(1);
                } else if (arr.get(i).get(0) == 2) {
                    ans += H-y + W-arr.get(i).get(1);
                }
            }
        }
        System.out.println(ans);
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}